### DISCUSSION TITLE
Documentation for multi-server deployment

### DISCUSSION THREAD

**Discussion by ian-bradbury** [bensheldon/good_job#211] [Q&A] [answered]
Is there any documentation or a blog/tutorial anywhere that shows how to deploy good_job onto a dedicated server?

I have questions in my head like......
Do I deploy the application source code to the dedicated server?
Do I also need still need to add the good_job gem to the application/web server?
How do I tell active job that I have a dedicated job server?

Any help and advice appreciated. 

Thanks.

**Comment by bensheldon** [comment, marked as answer]
Hey @ianbradbury-bizzibody. Great questions! And thanks for being the first user of the Github Discussion. This is an new feature for me 😄 

> Do I deploy the application source code to the dedicated server?

Yes. GoodJob is dependent on your entire Rails application stack. If you're creating a job-worker server, you would want to set it up exactly the same as your webservers (install your Rails applciation code, bundle gems, add environment variables, credentials, install binary dependencies like imagemagick). The only difference is that instead of running `bundle exec rails server` for the webserver. you'd have the main process be `bundle exec good_job start`

> Do I also need still need to add the good_job gem to the application/web server?

Yes. Your Rails gemfile should include `good_job`. These [set up instructions](https://github.com/bensheldon/good_job#set-up) are within the context of your Rails application. This is necessary so that when you enqueue a job with `perform_later`, it is sent to GoodJob.

> How do I tell active job that I have a dedicated job server?

As long as you've set your ActiveJob adapter to `:good_job`, nothing more is necessary (other than fine-tuning configuration). GoodJob communicates within itself through your Postgres database. When a job is enqueued on the webserver, it gets written to the database, and your jobworker will query that same database to find the job and execute it. 

Let me know what further questions you have. I'll say that infrastructure/deployment is a tough subject because everyone's situation and preferences are a little different.

**Reply by ian-bradbury** [reply]
@bensheldon - Thank you for this.  I now have a weekend project.  :)
